internal/inotify/inotify.go

```diff
@@ -0,0 +1,354 @@
+		paths:       map[int]string{},
+		watches:     map[string]int{},
+		namespaces:  map[string]struct{}{},
+	}
+	return n, n.WatchDir(root)
+}
+
+func (n *Notify) Start() {
+	n.WatchExistingLogs()
+	glog.V(0).Infoln("started....")
+	glog.Flush()
+	// Suggest to keep num of loops to 1, else events may be handled out of order
+	MaxEventLoops := 1
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go n.ReadLoop(&wg)
+	// TODO: remove the looping, need only one EventLoop goroutine
+	for i := 0; i < MaxEventLoops; i++ {
+		wg.Add(1)
+		go n.EventLoop(&wg, i)
+	}
+	wg.Wait()
+}
+
+// ReadLoop reads the inotify fd and generates events
+func (n *Notify) ReadLoop(wg *sync.WaitGroup) {
+	var (
+		buf [unix.SizeofInotifyEvent * EventChanSize]byte // Buffer for a maximum of 4096 raw events
+	)
+	defer func() {
+		close(n.events)
+		n.mtx.Lock()
+		for fd := range n.paths {
+			unix.InotifyRmWatch(n.fd, uint32(fd))
+		}
+		n.mtx.Unlock()
+		n.inotifyFile.Close()
+	}()
+
+	for {
+		readbytes, err := n.inotifyFile.Read(buf[:])
+		if err != nil {
+			if errors.Unwrap(err) == io.EOF {
+				glog.Errorf("Received EOF on inotify file descriptor")
+			}
+			glog.Errorf("Error in ReadLoop. breaking the loop. err: %v", err)
+			break
+		}
+		if readbytes <= 0 {
+			glog.Errorf("readbytes <= 0. breaking the loop. readbytes: %d", readbytes)
+			break
+		}
+		events := 0
+		consumed := 0
+		var offset uint32 = 0
+		for offset <= uint32(readbytes-unix.SizeofInotifyEvent) {
+			raw := (*unix.InotifyEvent)(unsafe.Pointer(&buf[offset]))
+			consumed += unix.SizeofInotifyEvent
+			path := string(buf[offset+unix.SizeofInotifyEvent : offset+unix.SizeofInotifyEvent+raw.Len])
+			consumed += int(raw.Len)
+			offset += unix.SizeofInotifyEvent + raw.Len
+			/*
+				if raw.Mask&unix.IN_IGNORED == unix.IN_IGNORED {
+					continue
+				}
+			*/
+			e := NotifyEvent{
+				InotifyEvent: *raw,
+				path:         strings.TrimRight(path, "\x00"),
+			}
+			n.events <- e
+			events += 1
+			atomic.AddUint64(&SentEvents, 1)
+		}
+		if readbytes-consumed != 0 {
+			glog.V(0).Infof("Read %d bytes, %d events, consumed %d bytes, remaining %d bytes", readbytes, events, consumed, (readbytes - consumed))
+		}
+	}
+	wg.Done()
+	glog.V(0).Infoln("exiting ReadLoop")
+}
+
+func (n *Notify) EventLoop(wg *sync.WaitGroup, idx int) {
+	var tickerCh <-chan time.Time
+	// start ticker in first goroutine only
+	if idx == 0 {
+		tickerCh = time.NewTicker(time.Minute * 1).C
+	}
+	for {
+		if glog.V(9) {
+			glog.Info("--------- going to select ---------")
+		}
+		select {
+		case e := <-n.events:
+			//handled := false
+			atomic.AddUint64(&HandledEvents, 1)
+			if e.IsOverFlowErr() {
+				glog.Exit("Overflow occured. Exiting program.")
+			}
+			n.mtx.RLock()
+			watchedPath, ok := n.paths[int(e.Wd)]
+			n.mtx.RUnlock()
+			if !ok {
+				glog.Errorf("A watch event received for an unknown watched path. Fd: %d, Event Mask: %x", e.Wd, e.Mask)
+				continue
+			}
+
+			if glog.V(5) && e.Mask != 0x8000 {
```

Comment:
Consider making the mask a constant so we  understanded what it represents

Reply:
ok